SimPEG/Regularization.py

```diff
@@ -61,320 +87,439 @@ def dim(self):
-    mesh    = None    #: A SimPEG.Mesh instance.
-    mref    = None    #: Reference model.
+    def __init__(
+        self, mesh=None, **kwargs
+    ):
 
-    def __init__(self, mesh=None, nP=None, mapping=None, indActive=None, **kwargs):
+        super(BaseRegularization, self).__init__()
+        self._mesh = mesh
         Utils.setKwargs(self, **kwargs)
-        assert isinstance(mesh, Mesh.BaseMesh), "mesh must be a SimPEG.Mesh object."
-        if indActive is not None and indActive.dtype != 'bool':
-            tmp = indActive
-            indActive = np.zeros(mesh.nC, dtype=bool)
-            indActive[tmp] = True
-        if indActive is not None and mapping is None:
-            mapping = Maps.IdentityMap(nP=indActive.nonzero()[0].size)
 
-        if mesh is None and nP is None:
-            raise Exception('either Mesh or number of parameters must be '
-                            'provided to the BaseRegularization')
-
-        self.regmesh = RegularizationMesh(mesh, indActive)
-        self.indActive = indActive
+    # Properties
+    mref = Props.SimPEGArray(
+        "reference model", default=Utils.Zero()
+    )
+    indActive = properties.Array(
+        "indices of active cells in the mesh", dtype=(bool, int)
+    )
+    cell_weights = properties.Array(
+        "regularization weights applied at cell centers", dtype=float
+    )
+
+    # Observers and Validators
+    @properties.validator('indActive')
+    def _cast_to_bool(self, change):
+        value = change['value']
+        if value is not None:
+            if value.dtype != 'bool':  # cast it to a bool otherwise
+                tmp = value
+                value = np.zeros(self.mesh.nC, dtype=bool)
+                value[tmp] = True
+                change['value'] = value
+
+        # update regmesh indActive
+        if getattr(self, '_regmesh', None) is not None:
+            self._regmesh.indActive = Utils.mkvc(value)
+
+    @properties.validator('mref')
+    def _validate_mref(self, change):
+        if not isinstance(change['value'], Utils.Zero) and self.nP != '*':
+            assert len(change['value']) == self.nP, (
+                'mref must be length {}'.format(self.nP)
+            )
+
+    @properties.validator('cell_weights')
+    def _validate_cell_weights(self, change):
+        if change['value'] is not None and self.nP != '*':
+            assert len(change['value']) == self.nP, (
+                'cell_weights must be length {}'.format(self.nP)
+            )
+
+    # Other properties and methods
+    @property
+    def nP(self):
+        """
+        number of model parameters
+        """
+        if getattr(self.mapping, 'nP') != '*':
+            return self.mapping.nP
+        elif getattr(self.regmesh, 'nC') != '*':
+            return self.regmesh.nC
+        else:
+            return '*'
 
-        if mesh is not None and nP is None:
-            nP = self.regmesh.nC
-        self.nP = nP
+    @property
+    def mesh(self):
```

Comment:
`Props.Mesh`? This could also have `allowed_mesh_types` or something. Which would be helpful...